File: services/operator/internal/registry/registry.go
package registry

import (
	"encoding/json"
	"sort"

	v1alpha1 "github.com/ray-provisioner/ray-crd-provisioner/services/operator/api/v1alpha1"
	"github.com/ray-provisioner/ray-crd-provisioner/services/operator/internal/raycluster"
)

const RegistryJSONKey = "plugins.json"

type PluginRecord struct {
	Name                string   `json:"name"`
	Namespace           string   `json:"namespace"`
	Enabled             bool     `json:"enabled"`
	Image               string   `json:"image"`
	Entrypoint          string   `json:"entrypoint"`
	DataTypes           []string `json:"dataTypes"`
	DependsOn           []string `json:"dependsOn"`
	SupportsDistributed bool     `json:"supportsDistributed"`
	WorkerGroupName     string   `json:"workerGroupName"`
	RayResourceName     string   `json:"rayResourceName"`
	ReadyReplicas       int32    `json:"readyReplicas"`
}

type registryDocument struct {
	Plugins []PluginRecord `json:"plugins"`
}

func BuildRegistryData(plugins []v1alpha1.RayPlugin) (map[string]string, error) {
	records := make([]PluginRecord, 0, len(plugins))
	for _, plugin := range plugins {
		workerGroupName := plugin.Status.WorkerGroupName
		if workerGroupName == "" {
			workerGroupName = raycluster.WorkerGroupName(plugin.Name)
		}
		records = append(records, PluginRecord{
			Name:                plugin.Name,
			Namespace:           plugin.Namespace,
			Enabled:             plugin.Spec.Enabled,
			Image:               plugin.Spec.Image,
			Entrypoint:          plugin.Spec.Plugin.Entrypoint,
			DataTypes:           append([]string(nil), plugin.Spec.Plugin.DataTypes...),
			DependsOn:           append([]string(nil), plugin.Spec.Plugin.DependsOn...),
			SupportsDistributed: plugin.Spec.Plugin.SupportsDistributed,
			WorkerGroupName:     workerGroupName,
			RayResourceName:     workerGroupName,
			ReadyReplicas:       plugin.Status.ReadyReplicas,
		})
	}

	sort.Slice(records, func(i, j int) bool {
		return records[i].Name < records[j].Name
	})

	payload, err := json.MarshalIndent(registryDocument{Plugins: records}, "", "  ")
	if err != nil {
		return nil, err
	}

	return map[string]string{RegistryJSONKey: string(payload)}, nil
}

File: packages/ray-framework/ray_framework/registry.py
import json
from dataclasses import dataclass, field
from typing import Any


@dataclass(frozen=True)
class PluginRecord:
    name: str
    enabled: bool
    data_types: list[str]
    depends_on: list[str] = field(default_factory=list)
    namespace: str = ""
    image: str = ""
    entrypoint: str = ""
    supports_distributed: bool = False
    worker_group_name: str = ""
    ray_resource_name: str = ""
    ready_replicas: int = 0


def load_registry_from_json(payload: str) -> list[PluginRecord]:
    document = json.loads(payload)
    return [_record_from_json(item) for item in document.get("plugins", [])]


def _record_from_json(item: dict[str, Any]) -> PluginRecord:
    return PluginRecord(
        name=item["name"],
        namespace=item.get("namespace", ""),
        enabled=item.get("enabled", False),
        image=item.get("image", ""),
        entrypoint=item.get("entrypoint", ""),
        data_types=list(item.get("dataTypes", [])),
        depends_on=list(item.get("dependsOn", [])),
        supports_distributed=item.get("supportsDistributed", False),
        worker_group_name=item.get("workerGroupName", ""),
        ray_resource_name=item.get("rayResourceName", ""),
        ready_replicas=item.get("readyReplicas", 0),
    )
